### PR TITLE
Templates restructure

### PR DESCRIPTION
**Pull request by andrewjschuang**
### <samp>🤖 Generated by Copilot at 2954825</samp>

This pull request improves the code generation logic and the quality of the generated code for Pipedream components. It refactors and simplifies some configuration and helper files, and adds new templates and variables for the language model prompts and system instructions. It also fixes a typo in the README file and passes the authentication code example to the language model.

### <samp>🤖 Generated by Copilot at 2954825</samp>

> _We're coding up a storm, me hearties, with templates and prompts_
> _We're fixing typos, refactoring, and adding some new docs_
> _We're passing `auth` to the language model, to make it generate right_
> _We're heaving on the ropes, me lads and lasses, on the count of three, heave-ho!_

### <samp>🤖 Generated by Copilot at 2954825</samp>

*  Refactor `generate_code` function to improve readability and performance ([link](https://github.com/PipedreamHQ/pipedream/pull/7816/files?diff=unified&w=0#diff-1072f68a9f44e052a79715bf3ead28286fa62e091d4df16711de9a55ba87ecc2L13-R22), [link](https://github.com/PipedreamHQ/pipedream/pull/7816/files?diff=unified&w=0#diff-1072f68a9f44e052a79715bf3ead28286fa62e091d4df16711de9a55ba87ecc2L34-R31), [link](https://github.com/PipedreamHQ/pipedream/pull/7816/files?diff=unified&w=0#diff-1072f68a9f44e052a79715bf3ead28286fa62e091d4df16711de9a55ba87ecc2L42-R44), [link](https://github.com/PipedreamHQ/pipedream/pull/7816/files?diff=unified&w=0#diff-1072f68a9f44e052a79715bf3ead28286fa62e091d4df16711de9a55ba87ecc2L58-R72), [link](https://github.com/PipedreamHQ/pipedream/pull/7816/files?diff=unified&w=0#diff-1072f68a9f44e052a79715bf3ead28286fa62e091d4df16711de9a55ba87ecc2L81-R85))
*  Refactor `get_env_var` function and `config` object to simplify configuration and avoid hardcoding ([link](https://github.com/PipedreamHQ/pipedream/pull/7816/files?diff=unified&w=0#diff-75d5a3dabdf1cce5000f903be0a7a4b5075a7c679f01e23adf164c8486e7167eL6-R27), [link](https://github.com/PipedreamHQ/pipedream/pull/7816/files?diff=unified&w=0#diff-75d5a3dabdf1cce5000f903be0a7a4b5075a7c679f01e23adf164c8486e7167eL46-R38))
*  Add `auth_example` parameter to `PipedreamOpenAPIAgent` class and `call_langchain` function to pass authentication code example to language model ([link](https://github.com/PipedreamHQ/pipedream/pull/7816/files?diff=unified&w=0#diff-53a92d1cb5d0f3bd57b82e0c495cc9c1966d4a5493ab88134930f6dbd93e0059R1-R3), [link](https://github.com/PipedreamHQ/pipedream/pull/7816/files?diff=unified&w=0#diff-53a92d1cb5d0f3bd57b82e0c495cc9c1966d4a5493ab88134930f6dbd93e0059L27-R48), [link](https://github.com/PipedreamHQ/pipedream/pull/7816/files?diff=unified&w=0#diff-53a92d1cb5d0f3bd57b82e0c495cc9c1966d4a5493ab88134930f6dbd93e0059L69-R98))
*  Add new files to `templates/actions` module to provide text for introduction, examples, and additional rules for generating action components ([link](https://github.com/PipedreamHQ/pipedream/pull/7816/files?diff=unified&w=0#diff-a02cc62d4388df7fdae8382f11d78c611fbbb97ec98e0edc2261f2b44aafe989R1-R96), [link](https://github.com/PipedreamHQ/pipedream/pull/7816/files?diff=unified&w=0#diff-a00c7a38399205f06acde15d8f6059ba01a36f746957b98756ed5b926011df1eR1), [link](https://github.com/PipedreamHQ/pipedream/pull/7816/files?diff=unified&w=0#diff-572371d6f752fc919a5413417e141a349401d6bb66799bd435b9d628b4f7cdcbR1-R12), [link](https://github.com/PipedreamHQ/pipedream/pull/7816/files?diff=unified&w=0#diff-2d4d16cbeaa810ebb19d7ac618bf7d2913dc83ddb3e58833e8e8003061faa7c7R1-R28), [link](https://github.com/PipedreamHQ/pipedream/pull/7816/files?diff=unified&w=0#diff-d0a0fe82bd451e571a186f59b2be01109c9c1d11ffe8cecbb5ff015104d1cbffR1-R69))
*  Add new files to `templates/apps` module to provide text for introduction, examples, auth, methods, and prop definitions for generating app code ([link](https://github.com/PipedreamHQ/pipedream/pull/7816/files?diff=unified&w=0#diff-c3806c8ea09ec4e77c4fda0c381de5f02b32a2980f893dc466b010c476622184R1-R9), [link](https://github.com/PipedreamHQ/pipedream/pull/7816/files?diff=unified&w=0#diff-c55328d613c535441593302d220742db3eac5b282e4ef9b1306def823d86ee18R1-R7), [link](https://github.com/PipedreamHQ/pipedream/pull/7816/files?diff=unified&w=0#diff-6dd09f48ba3aae69d1a4edde5af49a92da61f04fd9643df81d56464a6e6d3c7dR1-R6), [link](https://github.com/PipedreamHQ/pipedream/pull/7816/files?diff=unified&w=0#diff-b40fb37da463c59675161466e2c5d5df23b15519f3901ae1c3f4da2c8d652ae0R1-R136), [link](https://github.com/PipedreamHQ/pipedream/pull/7816/files?diff=unified&w=0#diff-daa87ba964d408b2afe777f53a35efcb6dc2eb5f5ecff8f50e4d3e69906c1cb3R1-R9), [link](https://github.com/PipedreamHQ/pipedream/pull/7816/files?diff=unified&w=0#diff-77e2508b30b39b70cb03fa0ff5e6503fe4492a58fa82c405666f28e38a35404cR1-R3))
*  Add new files to `templates/common` module to provide text for common explanations and instructions for generating component code ([link](https://github.com/PipedreamHQ/pipedream/pull/7816/files?diff=unified&w=0#diff-4586880b75834684fa792214bbd83145b38711fc47db6373a06603e22708c1e1R1-R15), [link](https://github.com/PipedreamHQ/pipedream/pull/7816/files?diff=unified&w=0#diff-bb732efed7a372e2f3f1424ad817d95fbab2246e18caf24eaa78d8ad123e938aR1-R63), [link](https://github.com/PipedreamHQ/pipedream/pull/7816/files?diff=unified&w=0#diff-4ce01f3078682366fca702bcb0100987ed2cbae4713eb75d94373b416710cde7R1-R7), [link](https://github.com/PipedreamHQ/pipedream/pull/7816/files?diff=unified&w=0#diff-eccc11522f1d3578a5754be08082ce0f371f77bc6004a1c6444f3e449aab315fR1-R49), [link](https://github.com/PipedreamHQ/pipedream/pull/7816/files?diff=unified&w=0#diff-3a4cd42491b8d5a401737c4f426c073ebf73952d59a5a7561d6688520eeedfacR1-R19))
*  Fix typo in README file ([link](https://github.com/PipedreamHQ/pipedream/pull/7816/files?diff=unified&w=0#diff-5babb188f06dcffe35daa5796466d08adb4c4d5682803d769609ea524a6750bdL42-R42))

### CODE REVIEW

File: packages/component_code_gen/code_gen/generate_component_code.py [hunk 1]
```diff
@@ -10,19 +10,16 @@
 
 
 def generate_code(app, prompt, templates, tries):
+    validate_inputs(app, prompt, templates, tries)
+    db = supabase_helpers.SupabaseConnector()
+    docs_meta = db.get_app_docs_meta(app)
+    auth_meta = db.get_app_auth_meta(app)
+    auth_example = f"Here's how authentication is done in {app}:\n{auth_meta['component_code_scaffold_raw']}"
```
Comment (dylburger) [inline packages/component_code_gen/code_gen/generate_component_code.py:17]: Generally I'd recommend two newlines separating these just for clarity (easier to read sections when you print out the full prompt for debugging)

Comment (andrewjschuang) [reply packages/component_code_gen/code_gen/generate_component_code.py]: Thanks, added more newlines

Comment (dylburger) [reply packages/component_code_gen/code_gen/generate_component_code.py]: Can you also handle the case where `component_code_scaffold_raw` isn't present in auth_meta? When I was working with Danny we hit that case (for new apps that hadn't synced with Supabase yet).

File: packages/component_code_gen/templates/actions/additional_rules.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
+additional_rules = """## Additional rules
+
+When you generate code, you must follow all of the rules above. Review the rules and think through them step-by-step before you generate code. Look at how these map to the example code and component API described above.
```
Comment (dylburger) [inline packages/component_code_gen/templates/actions/additional_rules.py:3]: There are some rules here that should be shared across actions and sources. I think you could move #1 - #5, and #9 and #10 to a common file, then keep the other action-specific rules here.

Maybe it make sense to have the "common rules" start with the header `## Rules`, then call this header here `## Rules for writing actions` (same thing for the source rules)?

**Review by dylburger — APPROVED** [review]

**Pull request merged**